aidial_adapter_bedrock/llm/model/claude/v3/adapter.py

```diff
@@ -108,17 +119,34 @@ async def invoke_streaming(
         async with self.client.messages.stream(
             messages=messages,
             model=self.model,
-            event_handler=UsageEventHandler,
             **params,
         ) as stream:
-            async for text in stream.text_stream:
-                consumer.append_content(text)
-            consumer.close_content(to_dial_finish_reason(stream.stop_reason))
+            prompt_tokens = 0
+            completion_tokens = 0
+            stop_reason = None
+            async for event in stream:
+                if isinstance(event, MessageStartEvent):
```

Comment:
Could you do match-case over the whole possible range of event types with a final exhaustiveness check via `assert_never`? so that we know nothing was missed.

```json
MessageStreamEvent = Union[
    TextEvent,
    InputJsonEvent,
    RawMessageStartEvent,
    RawMessageDeltaEvent,
    MessageStopEvent,
    RawContentBlockStartEvent,
    RawContentBlockDeltaEvent,
    ContentBlockStopEvent,
]
```

`InputJsonEvent`, `*ContentBlock*Event` - what are those for? When are they triggered?

Reply:
`InputJsonEvent` is used to stream part of input JSON for tools

Since need already full JSON, I ignore that events, and wait for `ToolUseBlock` event in `ContentBlockStopEvent`

Reply:
The problem with making full match/case for every event type, is that we would still just ingore most of them with something like

```
case  RawContentBlockStartEvent:
    pass
```

What's the value in that?

---

Also, Claude streaming is kinda weird, a lot of event are just duplicating each other, like that:

```python
RawMessageStartEvent(message=Message(id='msg_bdrk_019JdnE9G6pppJd8aNPQ9Jev', content=[], model='claude-3-sonnet-20240229', role='assistant', stop_reason=None, stop_sequence=None, type='message', usage=Usage(input_tokens=217, output_tokens=2)), type='message_start')
RawContentBlockStartEvent(content_block=TextBlock(text='', type='text'), index=0, type='content_block_start')
RawContentBlockDeltaEvent(delta=TextDelta(text='Okay', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text='Okay', snapshot='Okay')
RawContentBlockDeltaEvent(delta=TextDelta(text=',', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=',', snapshot='Okay,')
RawContentBlockDeltaEvent(delta=TextDelta(text=' let', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' let', snapshot='Okay, let')
RawContentBlockDeltaEvent(delta=TextDelta(text=' me', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' me', snapshot='Okay, let me')
RawContentBlockDeltaEvent(delta=TextDelta(text=' check', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' check', snapshot='Okay, let me check')
RawContentBlockDeltaEvent(delta=TextDelta(text=' the', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' the', snapshot='Okay, let me check the')
RawContentBlockDeltaEvent(delta=TextDelta(text=' weather', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' weather', snapshot='Okay, let me check the weather')
RawContentBlockDeltaEvent(delta=TextDelta(text=' for', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' for', snapshot='Okay, let me check the weather for')
RawContentBlockDeltaEvent(delta=TextDelta(text=' San', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' San', snapshot='Okay, let me check the weather for San')
RawContentBlockDeltaEvent(delta=TextDelta(text=' Francisco', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' Francisco', snapshot='Okay, let me check the weather for San Francisco')
RawContentBlockDeltaEvent(delta=TextDelta(text=' using', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' using', snapshot='Okay, let me check the weather for San Francisco using')
RawContentBlockDeltaEvent(delta=TextDelta(text=' the', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' the', snapshot='Okay, let me check the weather for San Francisco using the')
RawContentBlockDeltaEvent(delta=TextDelta(text=' provide', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=' provide', snapshot='Okay, let me check the weather for San Francisco using the provide')
RawContentBlockDeltaEvent(delta=TextDelta(text='d tool', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text='d tool', snapshot='Okay, let me check the weather for San Francisco using the provided tool')
RawContentBlockDeltaEvent(delta=TextDelta(text=':', type='text_delta'), index=0, type='content_block_delta')
TextEvent(type='text', text=':', snapshot='Okay, let me check the weather for San Francisco using the provided tool:')
ContentBlockStopEvent(index=0, type='content_block_stop', content_block=TextBlock(text='Okay, let me check the weather for San Francisco using the provided tool:', type='text'))
RawContentBlockStartEvent(content_block=ToolUseBlock(id='toolu_bdrk_01YA41GCapx4x46q82EXkBkV', input={}, name='get_weather', type='tool_use'), index=1, type='content_block_start')
RawContentBlockDeltaEvent(delta=InputJsonDelta(partial_json='', type='input_json_delta'), index=1, type='content_block_delta')
InputJsonEvent(type='input_json', partial_json='', snapshot={})
RawContentBlockDeltaEvent(delta=InputJsonDelta(partial_json='{"locat', type='input_json_delta'), index=1, type='content_block_delta')
InputJsonEvent(type='input_json', partial_json='{"locat', snapshot={})
RawContentBlockDeltaEvent(delta=InputJsonDelta(partial_json='io', type='input_json_delta'), index=1, type='content_block_delta')
InputJsonEvent(type='input_json', partial_json='io', snapshot={})
RawContentBlockDeltaEvent(delta=InputJsonDelta(partial_json='n": "San ', type='input_json_delta'), index=1, type='content_block_delta')
InputJsonEvent(type='input_json', partial_json='n": "San ', snapshot={})
RawContentBlockDeltaEvent(delta=InputJsonDelta(partial_json='Francisco"}', type='input_json_delta'), index=1, type='content_block_delta')
InputJsonEvent(type='input_json', partial_json='Francisco"}', snapshot={'location': 'San Francisco'})
ContentBlockStopEvent(index=1, type='content_block_stop', content_block=ToolUseBlock(id='toolu_bdrk_01YA41GCapx4x46q82EXkBkV', input={'location': 'San Francisco'}, name='get_weather', type='tool_use', __json_buf=b'{"location": "San Francisco"}'))
RawMessageDeltaEvent(delta=Delta(stop_reason='tool_use', stop_sequence=None), type='message_delta', usage=MessageDeltaUsage(output_tokens=71))
MessageStopEvent(type='message_stop', message=Message(id='msg_bdrk_019JdnE9G6pppJd8aNPQ9Jev', content=[TextBlock(text='Okay, let me check the weather for San Francisco using the provided tool:', type='text'), ToolUseBlock(id='toolu_bdrk_01YA41GCapx4x46q82EXkBkV', input={'location': 'San Francisco'}, name='get_weather', type='tool_use', __json_buf=b'{"location": "San Francisco"}')], model='claude-3-sonnet-20240229', role='assistant', stop_reason='tool_use', stop_sequence=None, type='message', usage=Usage(input_tokens=217, output_tokens=71)))
```

Reply:
You could compress all the other cases into a singe pattern:

```python
    case Foo() | Bar():
        pass
    case _:
        assert_never(event)  
```

This way we will notice if an new type of an event is introduced. Then we'll decide if we need to ignore it or handle in some way, instead of ignoring it silently.

Reply:
Fixed